runtime/domain.c

```diff
@@ -2096,6 +2115,111 @@ static void domain_terminate (void)
+  caml_memprof_delete_domain(domain_state);
+
+  /* We can not touch domain_self->interruptor after here
+     because it may be reused */
+  caml_remove_generational_global_root(&domain_state->dls_root);
+  caml_remove_generational_global_root(&domain_state->backtrace_last_exn);
+  caml_stat_free(domain_state->final_info);
+  caml_stat_free(domain_state->ephe_info);
+  caml_free_intern_state();
+  caml_free_extern_state();
+  /* FIXME: this fails the caml_mark_stack_is_empty assert in the debug runtime
+    caml_teardown_major_gc();
+   */
+
+  caml_finalise_heap();
+
+  caml_teardown_shared_heap(domain_state->shared_heap);
+  domain_state->shared_heap = 0;
+  caml_free_minor_tables(domain_state->minor_tables);
+  domain_state->minor_tables = 0;
+
+  caml_orphan_alloc_stats(domain_state);
+  /* Heap stats were orphaned by [caml_teardown_shared_heap] above.
+     At this point, the stats of the domain must be empty.
+
+     The sampled copy was also cleared by the minor collection(s)
+     performed above at [caml_empty_minor_heaps_once()], see the
+     termination-specific logic in [caml_collect_gc_stats_sample_stw].
+  */
+
+  /* TODO: can this ever be NULL? can we remove this check? */
+  if(domain_state->current_stack != NULL) {
+    caml_free_stack(domain_state->current_stack);
+  }
+  caml_free_backtrace_buffer(domain_state->backtrace_buffer);
+  caml_free_gc_regs_buckets(domain_state->gc_regs_buckets);
+}
+
+/* Try and terminate the currently running domain.
+   This is only invoked when extra domains are left running while the
+   main one is terminating. In this case, we are not in a state where
+   we can safely release resources. The best we can do is cancel the
+   extra running threads. */
+static void stw_terminate_domain(caml_domain_state *domain, void *data,
+  int participating_count,
+  caml_domain_state **participating)
+{
+  if (!pthread_equal(pthread_self(), *(pthread_t *)data)) {
+    if (caml_bt_is_self()) {
+      /* If this STW request is handled by the backup thread, the
+         domain thread is currently running C code. */
+      (void)pthread_cancel(domain_self->tid);
+      /* We are intentionally not waiting for the thread to terminate here,
+         and not decrementing the number of running domains either, since
+         we know we can't correctly clean up resources in this state. */
+    } else {
+      /* Domain threads forced to exit here will not have a chance to
+         run domain_terminate() on their own, so we need to ask the
+         backup thread to terminate here. */
+      terminate_backup_thread(domain_self);
+      /* No particular memory resource cleanup is attempted here, for we
+         have no idea which state each domain is in. */
+    }
+    pthread_exit(0);
+  }
+}
+
+void caml_stop_all_domains(void)
+{
+  atomic_store_relaxed(&domains_exiting, 1);
+
+  pthread_t myself = pthread_self();
+  caml_try_run_on_all_domains(stw_terminate_domain, &myself, NULL);
```

Comment:
We probably want to run this in a loop, because `_try_` indicates that it may do nothing and return `false` if it runs into a concurrent STW request from another domain -- see other use-cases, one may either loop until the call return `1` or check another variable that indicates that this domain or some other domain succeeded in starting a STW section that achieved the desired purpose.

Reply:
You're right. In a former version, this used to be part of a `while (!caml_domain_alone())` loop.